File: 21. MergeTwoSortedLists/MergeTwoSortedLists(1).go
// Runtime: 0 ms, faster than 100.00% of Go online submissions for Merge Two Sorted Lists.
// Memory Usage: 2.5 MB, less than 99.77% of Go online submissions for Merge Two Sorted Lists.
/**
* Definition for singly-linked list.
* type ListNode struct {
*     Val int
*     Next *ListNode
* }
*/
func mergeTwoLists(l1 *ListNode, l2 *ListNode) *ListNode {
    if (l1 == nil && l2 == nil){ return nil }
    if (l1 == nil) { return l2 }
    if (l2 == nil){ return l1 }

    var first, current *ListNode

    if (l1.Val < l2.Val){
        first = l1
        current = l1
        l1 = l1.Next
    }else{
        first = l2
        current = l2
        l2 = l2.Next
    }

    for (l1 != nil || l2 != nil){
        if (l1 == nil){
            current.Next = l2
            current = current.Next
            l2 = l2.Next
            continue
        }
        if (l2 == nil){
            current.Next = l1
            current = current.Next
            l1 = l1.Next
            continue
        }

        if (l1.Val < l2.Val){
            current.Next = l1
            current = current.Next
            l1 = l1.Next
            continue
        }else{
            current.Next = l2
            current = current.Next
            l2 = l2.Next
        }   
    }
    
    return first
}

File: 21. MergeTwoSortedLists/MergeTwoSortedLists(1).cs
// Runtime: 92 ms, faster than 86.07% of C# online submissions for Merge Two Sorted Lists.
// Memory Usage: 25.5 MB, less than 78.33% of C# online submissions for Merge Two Sorted Lists.
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public ListNode MergeTwoLists(ListNode l1, ListNode l2) {
        
        if (l1 == null && l2 == null)
            return null;
        if (l1 == null)
            return l2;
        if (l2 == null)
            return l1;
        
        ListNode first, current;
        
        if (l1.val < l2.val){
            first = current = l1;
            l1 = l1.next;
        }
        else{
            first = current = l2;
            l2 = l2.next;
        }   
        
        while (l1 != null || l2 != null){
            if (l1 == null){
                current.next = l2;
                current = current.next;
                l2 = l2.next;
                continue;
            }
            if (l2 == null){
                current.next = l1;
                current = current.next;
                l1 = l1.next;
                continue;
            }
            
            if (l1.val < l2.val){
                current.next = l1;
                current = current.next;
                l1 = l1.next;
                continue;
            }
            else{
                current.next = l2;
                current = current.next;
                l2 = l2.next;
            }   
        }
        
        return first;
    }
}